conans/model/graph_lock.py

```diff
@@ -2,297 +2,465 @@
+
+        :return: An ordered list of lists, each inner element is a tuple with the node ID and the
+                 reference (as string), possibly including revision, of the node
+        """
+        # First do a topological order by levels, the ids of the nodes are stored
+        levels = []
+        opened = list(self._nodes.keys())
+        while opened:
+            current_level = []
+            for o in opened:
+                node = self._nodes[o]
+                requires = node.requires
+                if node.python_requires:
+                    requires += node.python_requires
+                if node.build_requires:
+                    requires += node.build_requires
+                if not any(n in opened for n in requires):
+                    current_level.append(o)
+
+            current_level.sort()
+            levels.append(current_level)
+            # now initialize new level
+            opened = set(opened).difference(current_level)
+
+        # Now compute the list of list with prev=None, and prepare them with the right
+        # references to be used in cmd line
+        result = []
+        total_prefs = set()  # to remove duplicates, same pref shouldn't build twice
+        for level in levels:
+            new_level = []
+            for n in level:
+                locked_node = self._nodes[n]
+                if locked_node.prev is None and locked_node.package_id is not None:
+                    # Manipulate the ref so it can be used directly in install command
+                    ref = repr(locked_node.ref)
+                    if not self._revisions_enabled:
+                        if "@" not in ref:
+                            ref += "@"
+                    else:
+                        if "@" not in ref:
+                            ref = ref.replace("#", "@#")
+                    if ref not in total_prefs:
+                        new_level.append((n, ref))
+                        total_prefs.add(ref)
+            if new_level:
+                result.append(new_level)
+
+        return result
+
+    def complete_matching_prevs(self):
+        """ when a build_require that has the same ref and package_id is built, only one node
+        gets its PREV updated. This method fills the repeated nodes missing PREV to the same one.
+        The build-order only returned 1 node (matching ref:package_id).
+        """
+        groups = {}
+        for node in self._nodes.values():
+            groups.setdefault((node.ref, node.package_id), []).append(node)
+
+        for nodes in groups.values():
+            if len(nodes) > 1:
+                prevs = set(node.prev for node in nodes if node.prev)
+                if prevs:
+                    assert len(prevs) == 1, "packages in lockfile with different PREVs"
+                    prev = prevs.pop()
+                    for node in nodes:
+                        if node.prev is None:
+                            node.prev = prev
```

Comment:
I would avoid that extra `if` (probably no time diff)

```suggestion
                        node.prev = prev
```